### DISCUSSION TITLE
Unexpected output from APSW interactive shell

### DISCUSSION THREAD

**Discussion by ColMcp** [rogerbinns/apsw#538] [Q&A] [answered]
I ran a short Python script using `apsw` to populate a Sqlite database from a `utf-8` encoded file of SQL statements. When I use the `apsw` interactive shell to select a particular item from a table it shows `O Boto (B{LATIN SMALL LETTER O WITH CIRCUMFLEX}to)`. When I use macOS's default `sqlite3` shell to do the same selection I get `O Boto (Bôto)` (which is what I expected).

I recognise the text in the curly braces to be the Unicode name of the `ô` but what is it doing there?

**Comment by rogerbinns** [comment, marked as answer]
It is an intentional difference in functionality when the shell is outputting a formatted query results table.   If you look in [the example](https://rogerbinns.github.io/apsw/example.html#formatting-query-results-table) you can see different options for how strings are sanitized - the parameter is [described here](https://rogerbinns.github.io/apsw/ext.html#apsw.ext.format_query_table).  (The SQLite shell is generally happy to just dump whatever bytes it got as text straight out.)

If the apsw shell is started interactively (ie running with terminal input and output) then it defaults to [mode box](https://rogerbinns.github.io/apsw/shell.html#shell-cmd-mode).  There is also `qbox` where quoting values is turned on, and `table` which uses ASCII.

You can change the options with flags named like [format_query_table](https://rogerbinns.github.io/apsw/ext.html#apsw.ext.format_query_table) parameters:

    .mode box --string-sanitize 0 --truncate 17 --text_with 132

There is currently a bug showing help (eg `.mode box --help`)  That will be in the next release (in the next few weeks).

Separately from that I have been adding comprehensive [FTS5](https://www.sqlite.org/fts5.html) support to APSW which is also in the next release  That forced me to implement a whole bunch of Unicode stuff such as [text segmentation](https://www.unicode.org/reports/tr29/) and [line breaking](https://www.unicode.org/reports/tr14/), in addition to an up to date Unicode database because I needed information [unicodedata](https://docs.python.org/3/library/unicodedata.html) doesn't have.  That let me fix text widths in the console and line breaking in the table.  It handles everything Unicode can handle.

**Reply by ColMcp** [reply]
Thank you so much for the comprehensive reply, and for all the work you have put into `apsw` over the years!

**Reply by rogerbinns** [reply]
The bug was only showing the help/errors for the mode - the flags do work.  I'm glad you enjoy apsw.  The FTS5 stuff is really good, even if I do say so myself.
